### PR TITLE
Rewrite retry mechanism for gRPC write in gcsio

### PR DESCRIPTION
**Pull request by hongyegong**
This change mainly solves 2 issues:

1. After resumable write failed due to transient error, e.g. DEADLINE_EXCEED, committed size queried from server can be less than the size that has been uploaded thru client request. Current implementation doesn't handle this case. New implementation handles this by cached the past `numberOfRequestsToRetain` number of data chunk for later resume if needed.
2. This also fixes a bug of current retry mechanism that the bufferedInputStream could be reset to an invalid position

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 1]
```diff
@@ -179,7 +193,11 @@ private Object doResumableUpload() throws IOException {
       int retriesAttempted = 0;
       Hasher objectHasher = Hashing.crc32c().newHasher();
       do {
-        responseObserver = new InsertChunkResponseObserver(uploadId, writeOffset, objectHasher);
+        // Holds list of most recent number of numberOfRequestsToRetain requests, so upload can be
+        // rewound and re-sent upon non-transient errors.
+        List<InsertObjectRequest.Builder> requestList = new ArrayList<>();
```
Comment (mfschwartz) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: I realize this came from my original implementation but: ArrayList.remove is inefficient - requires copying elements. How about switching to a LinkedList?

Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: There are two commits on this PR. You must be looking at first commit not all commits. I actually figured that TreeMap will be a good fit here and updated the code accordingly.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 2]
```diff
@@ -294,49 +338,86 @@ public void run() {
...
+          for (int i = 0; i < requestList.size(); i++) {
+            // Take current request builder if the writerOffset of it is equal to the current
+            // writeOffset. Take the previous one if it's greater than current writeOffset. This
+            // will work as server will allow to re-upload a previously uploaded part of the object
+            // and ignore the bytes preceding the last committed offset.
+            InsertObjectRequest.Builder currentRequestBuilder = requestList.get(i);
+            if (currentRequestBuilder.getWriteOffset() == writeOffset) {
+              requestBuilder = currentRequestBuilder;
+              break;
+            } else if (currentRequestBuilder.getWriteOffset() > writeOffset && i > 0) {
```
[the remark this answers is not in the document]
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Please checkout the latest commit, the logic here changed quite a lot.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 3]
```diff
@@ -294,49 +338,86 @@ public void run() {
...
+      private InsertObjectRequest.Builder getSavedRequestBuilder(
+          InsertObjectRequest.Builder requestBuilder) {
+        // Resume will only work if the first request builder in the cache carries a offset not
+        // greater than the current writeOffset.
+        if (requestList.get(0).getWriteOffset() <= writeOffset) {
+          for (int i = 0; i < requestList.size(); i++) {
+            // Take current request builder if the writerOffset of it is equal to the current
+            // writeOffset. Take the previous one if it's greater than current writeOffset. This
+            // will work as server will allow to re-upload a previously uploaded part of the object
+            // and ignore the bytes preceding the last committed offset.
```
[the remark this answers is not in the document]
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Not sure it's feasible to trigger this particular condition during retry in integration env. I'll try to consolidate this in grpc doc/code.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 4]
```diff
@@ -294,49 +338,86 @@ public void run() {
...
+              break;
+            }
+          }
+        }
+        if (requestBuilder == null) {
+          nonTransientError =
+              new RuntimeException(
+                  String.format(
+                      "Didn't have enough data buffered for attempt to resume upload for uploadID"
+                          + " %s: last committed offset=%s, earliest buffered offset=%s.",
```
[the remark this answers is not in the document]
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Done.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 5]
```diff
@@ -294,49 +338,86 @@ public void run() {
...
 
-      public Object getResponse() throws IOException {
-        if (hasError()) {
-          throw getError();
+      // Handles the case when a writeOffset of data read previously is being processed. This
+      // happens if a transient failure happens while uploading, and can be resumed by querying the
+      // current committed offset.
+      private InsertObjectRequest.Builder getSavedRequestBuilder(
+          InsertObjectRequest.Builder requestBuilder) {
+        // Resume will only work if the first request builder in the cache carries a offset not
```
[the remark this answers is not in the document]
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Done.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 6]
```diff
@@ -294,49 +338,86 @@ public void run() {
...
             });
       }
 
-      public Object getResponse() throws IOException {
-        if (hasError()) {
-          throw getError();
+      // Handles the case when a writeOffset of data read previously is being processed. This
+      // happens if a transient failure happens while uploading, and can be resumed by querying the
+      // current committed offset.
+      private InsertObjectRequest.Builder getSavedRequestBuilder(
```
[the remark this answers is not in the document]
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: I actually already updated this method name to cachedRequest instead of savedRequest. Again PTAL all commits.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 7]
```diff
@@ -223,46 +225,88 @@ private Object doResumableUpload() throws IOException {
 
       private final long writeOffset;
       private final String uploadId;
+      private final List<InsertObjectRequest.Builder> requestList;
```
[the remark this answers is not in the document]
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: N/A

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 8]
```diff
@@ -223,46 +225,88 @@ private Object doResumableUpload() throws IOException {
...
+      private Throwable transientError = null;
+      // The last non-transient error to occur during the streaming RPC.
+      private Throwable nonTransientError = null;
       // The last error to occur during the streaming RPC. Present only on error.
       private IOException error;
       // The response from the server, populated at the end of a successful streaming RPC.
       private Object response;
       private ByteString chunkData;
       private Hasher objectHasher;
+      private boolean isReadyHandleExecuted = false;
```
[the remark this answers is not in the document]
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:240]: I'll just delete it, this was a line of code I added for a broken test.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 9]
```diff
@@ -179,7 +193,11 @@ private Object doResumableUpload() throws IOException {
       int retriesAttempted = 0;
       Hasher objectHasher = Hashing.crc32c().newHasher();
       do {
-        responseObserver = new InsertChunkResponseObserver(uploadId, writeOffset, objectHasher);
+        // Holds list of most recent number of numberOfRequestsToRetain requests, so upload can be
+        // rewound and re-sent upon non-transient errors.
+        List<InsertObjectRequest.Builder> requestList = new ArrayList<>();
+        responseObserver =
+            new InsertChunkResponseObserver(uploadId, writeOffset, objectHasher, requestList);
         // TODO(b/151184800): Implement per-message timeout, in addition to stream timeout.
```
[the remark this answers is not in the document]
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:201]: This is actually quite often seen in GCS connector code, usually go by `TODO(b/xxxxxxxx)`

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 10]
```diff
@@ -71,6 +77,14 @@
   // when writing to underlying channel raises error.
   private static final int UPLOAD_RETRIES = 10;
 
+  // Number of insert requests to retain, in case we need to rewind and resume an upload. Using too
+  // small of a number could risk being unable to resume the write if the resume point is an
+  // 'already-discarded buffer; and setting the value too high wastes RAM. Note: We could have a
```
[the remark this answers is not in the document]
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:81]: Done.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 11]
```diff
@@ -223,7 +225,12 @@ private Object doResumableUpload() throws IOException {
 
       private final long writeOffset;
       private final String uploadId;
+      private final TreeMap<Long, ByteString> dataChunkMap;
       private volatile boolean objectFinalized = false;
```
Comment (mfschwartz) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:229]: Why do we need volatile? Only a single thread operates in this class at a time, right?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 12]
```diff
@@ -223,7 +225,12 @@ private Object doResumableUpload() throws IOException {
 
       private final long writeOffset;
       private final String uploadId;
+      private final TreeMap<Long, ByteString> dataChunkMap;
```
Comment (mfschwartz) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:228]: Add a comment, "Maps from x to y ..." (i.e., what is the Long, what is the ByteString)?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 13]
```diff
@@ -294,49 +328,145 @@ public void run() {
...
+                                    uploadId, curWriteOffset));
+                      } else {
+                        retryMap.put(curWriteOffset, count + 1);
+                      }
+                      break;
+                    } else if (entry.getKey() > writeOffset) {
+                      chunkData = prevDataChunk;
+                      requestBuilder =
+                          buildRequestFromWriteOffsetAndDataChunk(prevWriteOffset, prevDataChunk);
+                      int count =
```
Comment (mfschwartz) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:382]: How about moving this check above the loop - it's loop-invariant, so doing that would make the loop shorter/more readable.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 14]
```diff
@@ -294,49 +328,145 @@ public void run() {
...
                   }
                 }
-                requestObserver.onNext(requestBuilder.build());
-
-                if (objectFinalized) {
-                  // Close the request half of the streaming RPC.
-                  requestObserver.onCompleted();
+                if (dataChunkMap.size() == numberOfRequestsToRetain) {
+                  dataChunkMap.remove(dataChunkMap.firstKey());
+                  retryMap.remove(dataChunkMap.firstKey());
```
Comment (mfschwartz) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:333]: Why do we need two different maps from offset to chunk?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 15]
```diff
@@ -294,49 +328,145 @@ public void run() {
...
+                  TreeMap<Long, ByteString> dataChunkMap) {
+                // Resume will only work if the first request builder in the cache carries an offset
+                // not greater than the current writeOffset.
+                InsertObjectRequest.Builder requestBuilder = null;
+                if (dataChunkMap.size() > 0 && dataChunkMap.firstKey() <= writeOffset) {
+                  long prevWriteOffset = 0;
+                  ByteString prevDataChunk = null;
+                  for (Map.Entry<Long, ByteString> entry : dataChunkMap.entrySet()) {
+                    // Take the previous entry if it's greater than current writeOffset.
+                    if (entry.getKey() == writeOffset) {
```
Comment (mfschwartz) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:361]: Why do we need to handle the case where entry.getKey() == writeOffset separately from the case where entry.getKey() > writeOffset? I think you just need to iterate until you get to the first entry whose offset is <= last persisted offset + write size (which means the cached request contains the offset where we want to resume writing). I think the code would be more readable if you have this one condition instead of two.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 16]
```diff
@@ -178,8 +192,12 @@ private Object doResumableUpload() throws IOException {
       long writeOffset = 0;
       int retriesAttempted = 0;
       Hasher objectHasher = Hashing.crc32c().newHasher();
+      // Holds list of most recent number of numberOfRequestsToRetain requests, so upload can be
+      // rewound and re-sent upon non-transient errors.
+      TreeMap<Long, ByteString> dataChunkMap = new TreeMap<>();
```
Comment (mfschwartz) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:197]: I assume the reason you chose a TreeMap is because you can then key on the offset. But since you always just iterate over it (rather than do a single lookup to find the needed chunk), I think this is overkill and kind of confusing; all you really need is a LinkedList, so you know you will iterate in offset order.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 13 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:384]: The count is actually related to iteration itself. We have no idea upfront which <offset, datachunk> will be retired and which are not until we iterate and find the first entry that meed resume requirement.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 15 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:363]: I am not sure I understand correctly "whose offset is <= last persisted offset + write size". I think regardless we need to diff btw '<' and '=='. So just to be clear, each entry in the map means <write offset, datachunk that will be written from this write offset>. So ideally the new writeOffset that comes with a new insert would be larger than all keys in the caching map. If not, there are two cases 1) new writeOffset is even smaller than the first key of the caching map, in which case we will just consider as non-resumable and as a in-transient error 2) this is the case we can resume from the caching map, if the new writeOffset is equal to any of previous key that means we can redo insertion with exactly that offset and the corresponding data chunk. However if there is no exact match, we want to find the first key(offset) in the map that is greater that this new offset but we do not use this one to redo insert because there will be piece of last data chunk that's gonna be missing(remember the current new writeOffset means committed size we queried from server). Instead we want to use the entry <offset, datachunk> one before the offset that's larger than new writeOffset. Hope this helps to explain...

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 14 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:335]: Second is keep track of how many tries have been made to each offset position. Otherwise the logic could end up trying a failed offset with associated data chunk endlessly.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 12 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:230]: Docstring added.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 16 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:198]: But we need to keep both value of offset, not just order, along with cached data chunk. Just the fact that data chunks are pushed in increasing offset order will not be enough to pick the right data chunk to resume write.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 11 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:231]: Not necessarily, for example the code in beforestart() will be running in a different thread and it could be we want the status of write is reflected in main memory once it's been marked as true in any thread.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 14 above)
Comment (mfschwartz) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:335]: This approach will potentially make us retry too many times (in the worst case we'll retry UPLOAD_RETRIES per chunk) and that the number of retries varies with object size being uploaded.

Let's please keep this simpler. Impose the UPLOAD_RETRIES limit at the top level, i.e., make that the maximum number of times we retry for a transient error. That will be easier to understand and consistent with the way retry handling is implemented in all other GCS clients. And it will let you get rid of this second map, making the code easier to read and reason about.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 16 above)
Comment (mfschwartz) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:198]: Oh, I see, your map no longer stores the pre-built request objects, but rather just the bytes (which is why you then also need to store the offset). In my implementation I held on to the request objects. The way you are doing it forces you to re-scan the data every time you re-transmit (to compute the checksums).

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 17]
```diff
@@ -64,12 +69,20 @@
   // Default GCS upload granularity.
   static final int GCS_MINIMUM_CHUNK_SIZE = 256 * 1024;
 
-  private static final Duration START_RESUMABLE_WRITE_TIMEOUT = Duration.ofMinutes(10);
-  private static final Duration QUERY_WRITE_STATUS_TIMEOUT = Duration.ofMinutes(10);
-  private static final Duration WRITE_STREAM_TIMEOUT = Duration.ofMinutes(20);
+  private static final Duration START_RESUMABLE_WRITE_TIMEOUT = Duration.ofSeconds(10);
+  private static final Duration QUERY_WRITE_STATUS_TIMEOUT = Duration.ofSeconds(10);
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:73]: Isn't this too low? Can we make is larger just to make sure that we are resilient to network issues? Maybe at least 1 minute.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 18]
```diff
@@ -156,6 +169,8 @@ public void startUpload(InputStream pipeSource) {
     // Read end of the pipe.
     private final BufferedInputStream pipeSource;
     private final int MAX_BYTES_PER_MESSAGE = MAX_WRITE_CHUNK_BYTES.getNumber();
+    // A map mapping from write offset to retries that have been made to the corresponding offset.
+    private final Map<Long, Integer> retryMap = new HashMap<>();
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:173]: Rename to `offsetRetryMap` and use `AtomicLong` as a value because you can increment it without changing value in the map.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 19]
```diff
@@ -223,7 +226,13 @@ private Object doResumableUpload() throws IOException {
 
       private final long writeOffset;
       private final String uploadId;
+      // A map holding cached data chunks, keyed by the offset to start writing each data chunk.
+      private final TreeMap<Long, ByteString> dataChunkMap;
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:230]: Why do we need 2 `dataChunkMap`s?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 20]
```diff
@@ -253,13 +267,35 @@ public void run() {
...
                 }
 
+                InsertObjectRequest.Builder requestBuilder = null;
+                // Create new request builder if the writeOffset grows over inserts. Otherwise
+                // something went wrong and resume write from a cached request builder.
+                if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
+                  requestBuilder = buildRequestFromCachedDataChunk(dataChunkMap);
+                } else {
+                  requestBuilder = buildRequestFromPipeData();
+                }
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:277]: Convert to ternary operator.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 21]
```diff
@@ -253,13 +267,35 @@ public void run() {
...
                 } catch (IOException e) {
-                  error =
-                      new IOException(
-                          String.format("Failed to read chunk for '%s'", resourceId), e);
-                  return;
+                  nonTransientError =
+                      new RuntimeException(
+                          String.format(
+                              "InsertChunkResponseObserver.beforeStart for uploadId %s caught %s",
+                              uploadId, e));
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:297]: Pass `e` as a parameter of the `RuntimeException` not `format` method, so its stack trace will be preserved.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 22]
```diff
@@ -294,49 +330,145 @@ public void run() {
...
               private ByteString readRequestData() throws IOException {
-                // Mark the input stream in case this request fails so that read can be recovered
-                // from where it's marked.
-                pipeSource.mark(MAX_BYTES_PER_MESSAGE);
                 ByteString data =
                     ByteString.readFrom(ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE));
-
-                objectFinalized =
-                    data.size() < MAX_BYTES_PER_MESSAGE || pipeSource.available() <= 0;
+                pipeSource.mark(1);
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:344]: Why do we mark with `1`?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 23]
```diff
@@ -178,8 +193,12 @@ private Object doResumableUpload() throws IOException {
       long writeOffset = 0;
       int retriesAttempted = 0;
       Hasher objectHasher = Hashing.crc32c().newHasher();
+      // Holds list of most recent number of numberOfRequestsToRetain requests, so upload can be
+      // rewound and re-sent upon non-transient errors.
+      TreeMap<Long, ByteString> dataChunkMap = new TreeMap<>();
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:198]: Are there any guarantees that `getCommittedWriteSize` will always return one of the previously sent offsets? I.e. is it always aligned to 2 MiB (chunk size)?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 24]
```diff
@@ -294,49 +330,145 @@ public void run() {
...
+              // querying the current committed offset.
+              private InsertObjectRequest.Builder buildRequestFromCachedDataChunk(
+                  TreeMap<Long, ByteString> dataChunkMap) {
+                // Resume will only work if the first request builder in the cache carries an offset
+                // not greater than the current writeOffset.
+                InsertObjectRequest.Builder requestBuilder = null;
+                if (dataChunkMap.size() > 0 && dataChunkMap.firstKey() <= writeOffset) {
+                  long prevWriteOffset = 0;
+                  ByteString prevDataChunk = null;
+                  for (Map.Entry<Long, ByteString> entry : dataChunkMap.entrySet()) {
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:361]: Because you already use TreeSet, may you use binary search methods to find offset that you need instead of iterating over them?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 25]
```diff
@@ -348,13 +480,28 @@ public void onNext(Object response) {
...
-                        + " Status: %s",
-                    resourceId, uploadId, writeOffset, statusDesc),
-                t);
+
+        if (t.getClass() == StatusException.class || t.getClass() == StatusRuntimeException.class) {
+          Code code =
+              t.getClass() == StatusException.class
+                  ? ((StatusException) t).getStatus().getCode()
+                  : ((StatusRuntimeException) t).getStatus().getCode();
+          if (code == Code.DEADLINE_EXCEEDED
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:489]: Create a `HashSet` of transient error codes and check if `code` is in this set instead of using many `||` statements.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 26]
```diff
@@ -348,13 +480,28 @@ public void onNext(Object response) {
...
+              || code == Code.RESOURCE_EXHAUSTED
+              || code == Code.INTERNAL
+              || code == Code.UNAVAILABLE) {
+            transientError = t;
+          }
+        }
+        if (transientError == null) {
+          nonTransientError =
+              new Throwable(
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:498]: Set `t` as a cause of this `Throwable` and change exception type of all thrown exceptions to `IOException` - I think this is more appropriate.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java [hunk 27]
```diff
@@ -64,12 +69,20 @@
...
-  private static final int UPLOAD_RETRIES = 10;
+  private static final int UPLOAD_RETRIES = 5;
+
+  // Number of insert requests to retain, in case we need to rewind and resume an upload. Using too
+  // small of a number could risk being unable to resume the write if the resume point is an
+  // already-discarded buffer; and setting the value too high wastes RAM. Note: We could have a
+  // more complex implementation that periodically queries the service to find out the last
+  // committed offset, to determine what's safe to discard, but that would also impose a performance
+  // penalty.
+  private static int numberOfRequestsToRetain = 5;
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:85]: Convert name to capitalized snake case, because this is a constant.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 19 above)
Comment (mfschwartz) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:230]: +1, I started a previous conversation about this.
I suggest we set up a meeting because we seem to be revisiting the same questions.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 24 above)
Comment (mfschwartz) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:361]: The list is only 5 long, so binary search is less efficient than just walking the list linearly.
I also previously suggested a TreeSet is overkill - all we need here is a simple LinkedList. Henry didn't want to do that because he wants to get the offset from the tree key but if you instead use as LinkedList of request messages (rather than just the byte strings) you have the offsets in the request messages and it has the other important advantage that you avoid re-scanning the full chunk data whenever re-transmitting.

Again, I think it's time for a meeting. I feel the current implementation is both overly complex and inefficient.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 17 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Updated. Although this is supposely not kind of requests that should take a minute before triggering retry.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 18 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Removed this map for now.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 19 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:240]: Discussed offline, removed the retryMap for now.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 20 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Done.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 21 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Done.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 22 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:356]: This is to revert to old approach to peek available bytes to avoid making a new request for nothing.
It was removed to avoid confusion because we relied on another reset to resume write, since we don't do that anymore hence revert to old way.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 23 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:200]: That's not assumed in the implementation, gRPC write updates with offsets like 2, 4, 6, 8Mibs and the size returned from server can be 7Mib, in that case we resume uploading from offset 6Mibs again and Michael confirmed that GCS is smart enough to start with 7 and drop the data from 6-7.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 24 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java:371]: Discussed offline. Will keep TreeSet and plain iteration for the sake of simplicity and readability.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 25 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Created a static Set for these codes for maintainability.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 26 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Done.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 27 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: Done.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 14 above)
Comment (hongyegong) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: As discussed offline, this map can be removed and will enforce the maximum retries of UPLOAD_RETRIES at top lvl.

**Comment by hongyegong** [comment]
> This PR reduces test coverage by 0.1%, please add test to cover untested functionality.

Added more test to cover more edge cases. Current unit test design is quite inflexible to add some test cases and needs a refactor.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java (hunk 17 above)
Comment (mfschwartz) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java]: It shouldn't take a minute, but 10 seconds was kind of short. You could use 30 seconds or just leave it at 1 min.
